pex/resolver.py

```diff
@@ -205,57 +207,64 @@ def resolve(requirements=None,
+  if not any((to_build, to_install)):
     # Nothing to build or install.
     return []
 
-  safe_mkdir(built_wheels)
-
   if to_build:
-    try:
-      build_wheels(distributions=to_build,
-                   target=built_wheels,
-                   cache=cache,
-                   interpreter=interpreter)
-    except PipError as e:
-      raise Untranslateable('Failed to build at least one of {}:\n\t{}'.format(to_build, str(e)))
-
-  if to_copy:
-    for wheel in to_copy:
-      dest = os.path.join(built_wheels, os.path.basename(wheel))
-      TRACER.log('Copying downloaded wheel from {} to {}'.format(wheel, dest))
-      shutil.copy(wheel, dest)
+    with TRACER.timed('Building distributions:\n  {}'.format('\n  '.join(to_build))):
+      try:
+        build_wheels(distributions=to_build,
+                     target=built_wheels,
+                     cache=cache,
+                     interpreter=interpreter)
+      except PipError as e:
+        raise Untranslateable('Failed to build at least one of {}:\n  {}'
+                              .format(', '.join(to_build), str(e)))
+      to_install.extend(os.path.join(built_wheels, wheel) for wheel in os.listdir(built_wheels))
 
   # 3. Chroot
   resolved_distributions = []
 
-  for wheel_file in os.listdir(built_wheels):
-    chroot = os.path.join(installed_chroots, wheel_file)
-    try:
-      install_wheel(wheel=os.path.join(built_wheels, wheel_file),
-                    target=chroot,
-                    compile=compile,
-                    overwrite=True,
-                    cache=cache,
-                    interpreter=interpreter)
-    except PipError as e:
-      raise Untranslateable('Failed to install {}:\n\t{}'.format(wheel_file, str(e)))
-
-    environment = Environment(search_path=[chroot])
-    for dist_project_name in environment:
-      resolved_distributions.extend(environment[dist_project_name])
-
-  markers_by_req_key = _calculate_dependency_markers(resolved_distributions,
-                                                     interpreter=interpreter)
+  with TRACER.timed('Installing distributions:\n  {}'.format('\n  '.join(to_install))):
+    for wheel_file_path in to_install:
+      wheel_file = os.path.basename(wheel_file_path)
+      chroot = os.path.join(installed_chroots, wheel_file)
+      if os.path.exists(chroot):
+        TRACER.log('Using cached installation of {} at {}'.format(wheel_file, chroot))
+      else:
+        TRACER.log('Installing {} in {}'.format(wheel_file_path, chroot))
+        tmp_chroot = '{}.{}'.format(chroot, uuid4().hex)
+        try:
+          install_wheel(wheel=wheel_file_path,
+                        target=tmp_chroot,
+                        compile=compile,
+                        overwrite=True,
+                        cache=cache,
+                        interpreter=interpreter)
+        except PipError as e:
+          raise Untranslateable('Failed to install {}:\n\t{}'.format(wheel_file_path, str(e)))
```

Comment:
The `str` has no effect. `{}` automatically calls `str()`.

```
In [3]: try:
   ...:     raise ValueError("doom")
   ...: except ValueError as e:
   ...:     print("{}".format(e))
   ...:
   ...:
doom
```

```
In [4]: try:
   ...:     raise ValueError("doom")
   ...: except ValueError as e:
   ...:     print("{}".format(str(e)))
   ...:
   ...:
doom
```

Also, consider using `!r` for the exception. I still believe that more information is useful, but not going to fight very hard for that one because I think you have a point that there's a downside.

Reply:
Will do on the str - I still completely disagree with the repr thing.